### ISSUE TITLE
Put 3D Secure authentication page in an iframe

### ISSUE THREAD

**Issue by offenb** [woocommerce/woocommerce-gateway-stripe#430]
It is common practice to put the 3d secure authentication page in an iframe so that the shopper has some re-assurance.

Many merchants also include reassurance messages around the iframe and also some back buttons in case there is a problem with the authentication (very occasionally the card issuer;s server could be down)

**Comment by roykho** [comment]
Unfortunately this is not controlled by us.

**Comment by offenb** [comment]
I did start by asking Stripe about this and they suggested it was the cart implementation that caused this (see below)....   I will go back to them


Question to Stripe:
You do 3D Secure authentication without using an iframe - this is unusual - or is this not stripe but the way our shopping cart has implemented it?

Stripe Answer:
3D Secure authentication is usually done using an iframe. This depends on the 3DS implementation of the shopping card you use, so the best course of action would be to get in touch with them.

**Comment by roykho** [comment]
Nope. Please read their documentation https://stripe.com/docs/sources/three-d-secure

**Comment by mdmoore** [comment]
@roykho Care to expand on this a bit more?  The docs only state this: 

`Your customer is then potentially redirected to their card issuer’s website to verify their identity using 3D Secure`

Is 3DS via iFrame not possible, or something that could be potentially added in the future?

**Comment by roykho** [comment]
@mdmoore not sure if it is possible or not because we can only follow what Stripe has capabilities for and in the documentation Stripe says a redirect will happen. See here ![](http://cld.wthms.co/5y8qgD+)

**Comment by thorsten-stripe** [comment]
it is technically possible, it just simply isn't a great experience across devices, especially on mobile. We're utilizing webhooks to catch users that drop off during the redirect and to be able to show an error upon redirect return if a customer aborted authentication.

Do you have data indicating that the redirect has a negative effect on conversion for your customers? Then we can reevaluate the integration path, otherwise, I'd recommend sticking with the homogenous redirect approach across all devices.

**Comment by offenb** [comment]
yes have used 3d secure with other gateways for years and a full page redirect spooks customers quite a lot for the following reasons

1. URL changes - customers don't understand the technology and a cardinal commerce url or other 3dsecure server url unsettles them

2. shopper not familiar with 3ds - with an iframe info about 3ds can be put on the page as many do not understand it

3. card issuer 3ds system down - as a shopper I have experienced card issuer 3ds page failing to load - when it is a framed page there are normally back buttons.  Full page redirect shopper may not know what to do

4. escape option for shopper - if they do not want to continue with 3ds the don't have to resort to browser back buttons

by not using iframe you are going against what i would say is latest best practice -  perhaps on mobile an iframe is not a good idea but on desktop it is - here is one example that includes info about 3ds, reassurance as the merchant's url still shows and an escape button.

<img width="1017" alt="screen shot 2018-01-16 at 10 48 21 am" src="https://user-images.githubusercontent.com/34250328/35006222-ef9fc7ce-faaa-11e7-8a8c-67050304c7e5.png">

**Comment by thorsten-stripe** [comment]
@offenb thanks, those points do make sense intuitively, but I'd love to get more data on the effect this actually has. We're currently evaluating this internally and might decide to build an iframe helper.

**Comment by offenb** [comment]
Your implementation is so archaic I thought I would have another quick look for more to support my request for iframe as you have re-implemented what was done more than 10 years ago when 3ds first started.

See page 47 
https://usa.visa.com/dam/VCOM/download/merchants/verified-by-visa-acquirer-merchant-implementation-guide.pdf

See page 3-8
http://arm-pdf-directory.s3.amazonaws.com/SMI_ManualMerchant.pdf

You will see both of these guides instruct the use of iframe and not what you have implemented. 

Mastercard even say "one approach, inline windows, is now acceptable for deployment. and existing merchants are expected to convert to an inline window implementation."

**Comment by thorsten-stripe** [comment]
Again, this recommends a framed solution stating the same reason, potential drop-off, but doesn't dictate it. We're working with the card networks on improving 3D Secure in general[0] and in light of that work will probably see a rework of this on the basis of 3D Secure V2 later this year anyway.

[0] https://usa.visa.com/visa-everywhere/security/future-of-digital-payment-security.html

**Comment by offenb** [comment]
It does dictate it.   

Visa doc says:

"The 3-D Secure Protocol **requires the authentication page** displays to be presented to
cardholders using the full browser window in an “inline” approach. U.S. merchant 3-D Secure
implementations **must use a framed inline page** for Verified by Visa. "

MasterCard says:

"There are two primary methods for creation of this window, however **only one approach,
inline windows, is now acceptable for deployment**."

**Comment by johndcoy** [comment]
Reopening and labeling for enhancement.

**Comment by offenb** [comment]
Disappointing that this is still left as low priority.   

We have multiple instances of customers being re-directed to a card issuers 3DS server and the page not loading (for whatever reason).   They just end up with an empty browser and don't really know what has happened, some think they placed an order, others realise they haven't, other... well who knows what conclusion they come up with.

I will remind you your implementation does not meet the specifications of the card networks.

I attach another implementation from a large retailer - if their shopper faces this issue, the end up with just part of the page not loading, the can see from the progress bar they are not at the end and they have some escape/recovery links at the bottom of the page

![image](https://user-images.githubusercontent.com/34250328/41463819-c457395c-708f-11e8-83b7-e4492f665641.png)
